### PR TITLE
Fixes #15846 - Renamed use_puppet_default to skip_foreman in LookupValue.

### PR DESCRIPTION
**Pull request by ShimShtein**
As part of this commit added a helper concern for deprecation of API properties.
## Usage

In the controller include `Api::V2::ParameterDeprecation`, and call `deprecate_param`. 
Example:

``` ruby
class OverrideValuesController < V2::BaseController
  include Api::V2::ParameterDeprecation

  deprecate_param [:create, :update], [:override_value, :use_puppet_default], :skip_foreman
end
```

`deprecate_param` parameters:
- action: (array of symbols) What actions are affected by the deprecation
- property_path: (array of symbols) What is the path to the deprecated property in `params` hash. In our example, it represents the property `params[:override_value][:use_puppet_default]`
- new_name: (nillable string) If a property was not dropped but renamed, set this value to the new name, and it will be filled automatically by the framework.

### CODE REVIEW

**Comment by ShimShtein** [comment]
Rebased, retesting.
[test]

**Comment by ShimShtein** [comment]
Rebased

File: app/controllers/concerns/api/v2/parameter_deprecation.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# This module adds a pre-action filter that will handle deprecated properties.
+# It can either just warn that the property is deprecated, in case there is no replacement for the property.
+# In case that a property was renamed, it will set the value of the new property to the value of the old one
+# in addition to the detailed warning.
+# Usage:
+# include this module, and mark a property as deprecated by using deprecate_param method.
+module Api::V2::ParameterDeprecation
```
Comment (orrabin) [inline app/controllers/concerns/api/v2/parameter_deprecation.rb]: `use_puppet_default` is already in the api so I'd suggest keeping it as an alias and changing in v3 but I don't see reason to add a warning that we are going to deprecate it if we don't have plans on changing it in v2.
If there is need to deprecate in v2 maybe add a separate issue for this? then it's possible to decide what should be deprecated and when.

**Comment by ShimShtein** [comment]
Removed warning from deprecated properties, and set the new value as required. Now you can only rename properties, but not get rid of them completely.

File: app/helpers/common_parameters_helper.rb [hunk 2]
```diff
@@ -19,8 +19,8 @@ def parameter_value_content(id, value, options)
     content_tag(:span, options[:popover], :class => "input-group-addon") + lookup_key_field(id, value, options)
   end
 
-  def use_puppet_default_help link_title = nil, title = _("Use Puppet default")
-    popover(link_title, _("Do not send this parameter via the ENC.<br>Puppet will use the value defined in the manifest."), :title => title)
+  def skip_foreman_help link_title = nil, title = _("Skip prameter")
```
Comment (orrabin) [inline app/helpers/common_parameters_helper.rb]: Now that the default title is more generic please change the places that use this helper and add a specific title to them.

File: app/controllers/api/v2/override_values_controller.rb [hunk 3]
```diff
@@ -11,6 +12,8 @@ class OverrideValuesController < V2::BaseController
       before_action :return_if_smart_mismatch, :only => [:index, :create, :show, :update, :destroy]
       before_action :return_if_override_mismatch, :only => [:show, :update, :destroy]
 
+      deprecate_param [:create, :update], [:override_value, :use_puppet_default], :skip_foreman
```
Comment (orrabin) [inline app/controllers/api/v2/override_values_controller.rb]: Why is this still here if we are not deprecating it for now?

File: app/helpers/common_parameters_helper.rb [hunk 4]
```diff
@@ -19,8 +19,8 @@ def parameter_value_content(id, value, options)
...
+    popover(link_title, _("Foreman will not manage this parameter."), :title => title)
```
Comment (orrabin) [inline app/helpers/common_parameters_helper.rb]: Maybe the body should also have an option to override?

File: app/models/concerns/puppet_lookup_value_extensions.rb [hunk 5]
```diff
@@ -0,0 +1,11 @@
+module PuppetLookupValueExtensions
```
Comment (orrabin) [inline app/models/concerns/puppet_lookup_value_extensions.rb:1]: Is this in a concern because you want to later move it to a plugin?

File: app/controllers/concerns/api/v2/parameter_deprecation.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
+# This module adds a pre-action filter that will handle deprecated properties.
```
Comment (tbrisker) [inline app/controllers/concerns/api/v2/parameter_deprecation.rb]: Why not just use something like https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/template_combinations_controller.rb#L81 instead of this whole complicated module? API deprecations should be a very rare event, as the API is considered stable, and should only be done when there is no other option.

File: app/controllers/api/v2/override_values_controller.rb (hunk 3 above)
Comment (ShimShtein) [inline app/controllers/api/v2/override_values_controller.rb]: We do want to use the new parameter in code but we are not changing the API, so we need to rename the parameter before it is processed by the controller.

File: app/models/concerns/puppet_lookup_value_extensions.rb (hunk 5 above)
Comment (ShimShtein) [inline app/models/concerns/puppet_lookup_value_extensions.rb:1]: Yes.

File: app/controllers/api/v2/override_values_controller.rb (hunk 3 above)
Comment (orrabin) [inline app/controllers/api/v2/override_values_controller.rb]: There are already a few places that do that without `deprecate_param`.
Maybe move the deprecation to a separate pr and use it in more than one place?

File: app/controllers/concerns/api/v2/parameter_deprecation.rb (hunk 6 above)
Comment (ShimShtein) [inline app/controllers/concerns/api/v2/parameter_deprecation.rb]: I was not aware of that deprecation at all. Now that we already have more than one deprecation, we definitely should DRY it up. IMHO it will give us a single place to manage such things, which will ease up finding these places later on.
I am not sure that the module introduces that much complicity as opposed to dedicated method deprecation.

I think that uniformity of the handling takes place here. A good example is the deprecation warning - in my previous version, I was sending a warning message, but @orrabin asked me to remove the warning.

File: app/controllers/api/v2/override_values_controller.rb (hunk 3 above)
Comment (tbrisker) [inline app/controllers/api/v2/override_values_controller.rb]: I'm against introducing a 60 line, very complex module, for something that should not occur more then a handful of times at most and can be done in 5 easy to read and very explicit lines, even if they repeat 5 times in the code base. Readability in my opinion is more important here then then DRYing the code.

File: app/controllers/concerns/api/v2/parameter_deprecation.rb (hunk 6 above)
Comment (tbrisker) [inline app/controllers/concerns/api/v2/parameter_deprecation.rb]: I disagree, I don't think I'm a novice rails developer and I'm having a hard time understanding this module. We already have a standard mechanism for deprecation warnings as seen in the linked example - `Foreman::Deprecation.api_deprecation_warning`.

Comment (ShimShtein) [inline app/controllers/concerns/api/v2/parameter_deprecation.rb]: What about simplifying the module, would it be enough?

Most of the complication comes from traversing the `params` tree (`params[:a][:b][:c]`). I can "outsource" it as lambdas in the declaration. Would it be better?

``` ruby
deprecate_param [:create, :update], ->(params) { params[:override_value][:use_puppet_default] }, ->(params, value) { params[:override_value][:skip_foreman] = value }
```

Comment (tbrisker) [inline app/controllers/concerns/api/v2/parameter_deprecation.rb]: That's even more complicated, why do you need to traverse a tree to rename a parameter? what's wrong with something along the lines of this, if you insist on generalizing this:

``` ruby
   def deprecate_rename_param(old,new)
     if !params[old].nil?
       params[new] = params.delete(old)
       Foreman::Deprecation.api_deprecation_warning("#{old} was renamed to #{new}")
     end
   end
```

Comment (ShimShtein) [inline app/controllers/concerns/api/v2/parameter_deprecation.rb]: In our case the parameter is not on the first level: it's `params[:override_values][:use_puppet_default]`.

Comment (tbrisker) [inline app/controllers/concerns/api/v2/parameter_deprecation.rb]: Sounds to me like a good reason to go with a 5 line handler that may have something in common with other cases, instead of trying to force a complex generalization.

**Comment by ShimShtein** [comment]
Removed override module and moved body to a parameter in helper.

**Comment by orrabin** [comment]
[test]

File: app/helpers/lookup_keys_helper.rb [hunk 7]
```diff
@@ -118,7 +118,7 @@ def lookup_key_warnings(required, has_value)
       { :text => _("Required parameter without value.<br/><b>Please override!</b><br/>"),
         :icon => "error-circle-o" }
     else
-      { :text => _("Optional parameter without value.<br/><i>Will not be sent to Puppet.</i><br/>"),
+      { :text => _("Optional parameter without value.<br/><i>Still managed by Foreman, the value will be empty.</i><br/>"),
```
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb:121]: This change is not quite clear and the meaning seems to be different then before - empty value sent instead of not sent at all?

File: app/helpers/lookup_keys_helper.rb [hunk 8]
```diff
@@ -153,15 +153,15 @@ def lookup_value(host_or_hostgroup, lookup_key)
...
-    check_box(lookup_value_name_prefix(lookup_key.id), :use_puppet_default,
+    check_box(lookup_value_name_prefix(lookup_key.id), :skip_foreman,
               :value    => lookup_value.id,
               :disabled => disabled || !can_edit_params?,
-              :onchange => "toggleUsePuppetDefaultValue(this, 'value')",
+              :onchange => "toggleSkipForemanValue(this, 'value')",
               :hidden   => disabled,
-              :title    => _('Use Puppet default'),
-              :checked  => lookup_value.use_puppet_default)
+              :title    => _('Skip from ENC'),
```
Comment (tbrisker) [inline app/helpers/lookup_keys_helper.rb]: What does `Skip from ENC` mean? it's not clear

File: app/views/lookup_keys/_values.html.erb [hunk 9]
```diff
@@ -8,8 +8,8 @@
       </span></th>
     <% if is_param %>
       <th class='col-md-4'><%= _('Value') %></th>
-      <th class='col-md-2 ca'><%= _('Use Puppet default') %>
-        <span class="help-inline"> <%= use_puppet_default_help() %></span>
+      <th class='col-md-2 ca'><%= _('Skip parameter') %>
```
Comment (tbrisker) [inline app/views/lookup_keys/_values.html.erb]: Again - not clear what it means to skip the parameter?

File: app/views/puppetclasses/_classes_parameters.html.erb [hunk 10]
```diff
@@ -4,7 +4,7 @@
       <th class='col-md-3'><%= _('Puppet class') %></th>
       <th class='col-md-2'><%= _('Name') %></th>
       <th class='col-md-6'><%= _('Value') %></th>
-      <th class='col-md-1 ca'><%= _("Use Puppet default") %>&nbsp;<%= use_puppet_default_help %></th>
+      <th class='col-md-1 ca'><%= _("Skip from ENC") %>&nbsp;<%= skip_foreman_help %></th>
```
Comment (tbrisker) [inline app/views/puppetclasses/_classes_parameters.html.erb]: ditto

**Comment by tbrisker** [comment]
Other then a couple of inline comments about clarity of text looks fine to me.

File: app/helpers/lookup_keys_helper.rb (hunk 7 above)
Comment (ShimShtein) [reply app/helpers/lookup_keys_helper.rb:121]: In this case empty value will be sent:
image

Will lead to this output in the enc:

```
  XO4txkfDEUD: ''
```

**Comment by ShimShtein** [comment]
Changed the texts to be more clear and consistent.

File: app/assets/javascripts/lookup_keys.js [hunk 11]
```diff
@@ -140,14 +140,14 @@ function undo_remove_child_node(item){
 function toggleOverrideValue(item) {
   var override = $(item).is(':checked');
   var fields = $(item).closest('.fields');
-  var fields_to_disable = fields.find("[name$='[required]'],[id$='_key_type'],[id$='_validator_type'],[name$='[use_puppet_default]'],[name$='[hidden_value]']");
-  var use_puppet_default = $(item).closest('fieldset').find("[id$='use_puppet_default']").is(':checked');
+  var fields_to_disable = fields.find("[name$='[required]'],[id$='_key_type'],[id$='_validator_type'],[name$='[skip_foreman]'],[name$='[hidden_value]']");
```
Comment (orrabin) [inline app/assets/javascripts/lookup_keys.js]: Maybe change the name everywhere to omit and not only in the UI?

**Comment by ShimShtein** [comment]
Renamed the field to `omit` too.

**Comment by ShimShtein** [comment]
Renamed the JS function and the comment @tbrisker pointed out. No more "skip" in the code :)

**Comment by orrabin** [comment]
I know you didn't change use puppet default on default value because it will change when this pr is merged: https://github.com/theforeman/foreman/pull/3698 but that means there is no option to even test this pr until 3698 is merged.

It doesn't disable the value for default value:
image

and when I check override I can't check use puppet default until I save and edit the puppet class again.

Either leave this pr on WoC until the other is merged and then rebase or you need to change it here because having a mix of use puppet default and omit is confusing.

**Comment by tbrisker** [comment]
+1 to changing the default value one as well here so we aren't blocked on the other PR - if i'm not mistaken that should be a trivial change

**Comment by ShimShtein** [comment]
No more `use_puppet_default` across the board. @orrabin, this fixed the bug that you have described too (at least on my machine).

**Comment by ShimShtein** [comment]
Rebased

**Comment by orrabin** [comment]
@ShimShtein updating use_puppet_default through the api for a smart_class_parameter results in the error `unknown attribute 'use_puppet_default' for PuppetclassLookupKey`.
This should still work in v2 with a deprecation warning.

Edit: please also add a functional test for this, they only cover the new `omit` option.

**Comment by ShimShtein** [comment]
Fixed error and test added, thanks @orrabin !

**Comment by tbrisker** [comment]
@ShimShtein The commit message also needs to be updated

**Comment by ShimShtein** [comment]
Changed the commit message.

File: app/controllers/api/v2/override_values_controller.rb [hunk 12]
```diff
@@ -94,6 +97,17 @@ def return_if_override_mismatch
       def resource_class
         LookupValue
       end
+
+      def rename_use_puppet_default
+        return unless params[:override_value]
+
+        old_value = params[:override_value].delete(:use_puppet_default)
+
+        if old_value.present?
```
Comment (orrabin) [inline app/controllers/api/v2/override_values_controller.rb]: `use_puppet_default` is a boolean and when it is false `old_value.present?` returns false and the value doesn't get changed.

File: test/functional/api/v2/smart_class_parameters_controller_test.rb [hunk 13]
```diff
@@ -159,6 +159,15 @@ class Api::V2::SmartClassParametersControllerTest < ActionController::TestCase
     refute_equal orig_value, new_value
   end
 
+  test "should update smart class parameter with use_puppet_default (compatibility test)" do
+    orig_value = lookup_keys(:five).omit
+    refute lookup_keys(:five).omit # check that the initial value is false
+    put :update, { :id => lookup_keys(:five).to_param, :smart_class_parameter => { :use_puppet_default => "true" } }
```
Comment (orrabin) [inline test/functional/api/v2/smart_class_parameters_controller_test.rb:165]: please also add a test that check for false due to my comment above.

File: app/controllers/api/v2/override_values_controller.rb (hunk 12 above)
Comment (tbrisker) [reply app/controllers/api/v2/override_values_controller.rb]: how about:

``` ruby
return unless params[:override_value] && params[:override_value].has_key(:use_puppet_default)
params[:override_value][:omit] = params[:override_value].delete(:use_puppet_default)
Foreman::Deprecation.api_deprecation_warning('"use_puppet_default" was renamed to "omit"')
```

**Comment by ShimShtein** [comment]
Tests added, `present?` eliminated.

**Comment by orrabin** [comment]
@ShimShtein I tested and it looks good

File: app/controllers/api/v2/smart_class_parameters_controller.rb [hunk 14]
```diff
@@ -39,7 +41,8 @@ def show
         param :description, String, :desc => N_("Description of smart class")
         param :default_value, String, :desc => N_("Value to use when there is no match")
         param :hidden_value, :bool, :desc => N_("When enabled the parameter is hidden in the UI")
-        param :use_puppet_default, :bool, :desc => N_("Do not send this parameter via the ENC. Puppet will use the value defined in the Puppet manifest for this parameter")
+        param :use_puppet_default, :bool, :desc => N_("Deprecated, please use omit")
+        param :omit, :bool, :desc => N_("Do not send this parameter via the ENC. Puppet will use the value defined in the Puppet manifest for this parameter")
```
Comment (tbrisker) [inline app/controllers/api/v2/smart_class_parameters_controller.rb]: Shouldn't this (and in the other api controller) be changed to something like "omit from classification output" in line with the UI description?

File: app/controllers/concerns/foreman/controller/parameters/lookup_key.rb [hunk 15]
```diff
@@ -18,6 +18,7 @@ def add_lookup_key_params_filter(filter)
         :path,
         :puppetclass_id,
         :use_puppet_default,
```
Comment (tbrisker) [inline app/controllers/concerns/foreman/controller/parameters/lookup_key.rb]: shouldn't this also be removed?

**Comment by tbrisker** [comment]
:+1: other then a couple of minor comments inline.

**Comment by ShimShtein** [comment]
Done!

File: app/controllers/api/v2/smart_class_parameters_controller.rb [hunk 16]
```diff
@@ -39,7 +41,8 @@ def show
         param :description, String, :desc => N_("Description of smart class")
         param :default_value, String, :desc => N_("Value to use when there is no match")
         param :hidden_value, :bool, :desc => N_("When enabled the parameter is hidden in the UI")
-        param :use_puppet_default, :bool, :desc => N_("Do not send this parameter via the ENC. Puppet will use the value defined in the Puppet manifest for this parameter")
+        param :use_puppet_default, :bool, :desc => N_("Deprecated, please use omit")
+        param :omit, :bool, :desc => N_("Foreman will not send this parameter in classification output. Puppet will use the value defined in the Puppet manifest for this parameter")
```
Comment (tbrisker) [inline app/controllers/api/v2/smart_class_parameters_controller.rb:45]: Still have puppet here, isn't the point to make this CM-agnostic?

File: app/helpers/common_parameters_helper.rb [hunk 17]
```diff
@@ -19,8 +19,12 @@ def parameter_value_content(id, value, options)
     content_tag(:span, options[:popover], :class => "input-group-addon") + lookup_key_field(id, value, options)
   end
 
-  def use_puppet_default_help link_title = nil, title = _("Use Puppet default")
-    popover(link_title, _("Do not send this parameter via the ENC.<br>Puppet will use the value defined in the manifest."), :title => title)
+  def omit_help(
+      link_title = nil,
+      title = _("Omit prameter from classification"),
+      body = _("Foreman will not send this parameter in classification output.")
```
Comment (tbrisker) [inline app/helpers/common_parameters_helper.rb]: any reason to add this as a function parameter? Or to have any parameters at all since they are always used with the default values?

File: app/controllers/api/v2/smart_class_parameters_controller.rb (hunk 16 above)
Comment (orrabin) [reply app/controllers/api/v2/smart_class_parameters_controller.rb:45]: The `omit` in general - yes but since this will be just in the api docs for smart_class_parameter and not anywhere generic I'm find with keeping it a little puppet aware even just to not confuse people that are used to `use_puppet_default`.
I didn't comment on the UI helper because we want to reuse it with the same text for smart variables and that actually uses most of the same view but has a separate api controller.

File: app/helpers/common_parameters_helper.rb (hunk 17 above)
Comment (orrabin) [reply app/helpers/common_parameters_helper.rb]: +1 for removing the parameters because of my comment above, we keep it generic for reuse so it doesn't make sense to have it.
The only possible reason to leave it is if you do want to change the text per view.

Comment (tbrisker) [reply app/helpers/common_parameters_helper.rb]: I would suggest adding parameter if and when we need them, no point having parameters with defaults if we always use defaults.

Comment (orrabin) [reply app/helpers/common_parameters_helper.rb]: I meant change them now, have this generic body but send a puppet specific one in the view.

**Comment by ShimShtein** [comment]
I'll stick with the simpler solution - I'll remove parameters from the helper, since we are always use defaults. We can add the detailed messages later, if it will be requested.

**Comment by tbrisker** [comment]
looks fine to me. @orrabin - do you have other comments? if not feel free to merge once tests are green.

**Comment by ShimShtein** [comment]
[test foreman]

**Comment by orrabin** [comment]
merged as 569650b7d397bfd020dec53194a78f1c3813045f, thanks @ShimShtein
